### PR TITLE
[API] Validate RoleSet role specifications

### LINKED ISSUE #2583: RoleSet and StormService should reject invalid role specs during admission
### Bug description

While smoke-testing the latest upstream `main` controller and CRDs on minikube, I found several RoleSet / StormService spec validation gaps that can produce misleading Ready status or a repeated controller reconcile loop.

The most important cases are:

1. Duplicate `RoleSet.spec.roles[*].name` is accepted.
   - Observed: only one Pod is created for the duplicated role name, but RoleSet status contains two entries with the same role name and both are reported ready.
   - Impact: status aggregation becomes misleading and can confuse users or higher-level automation.

2. Invalid role names such as `bad_role` are accepted.
   - Observed: the controller tries to create Pods with names derived from the invalid role name, Kubernetes rejects those Pod names, and the RoleSet is reconciled repeatedly.
   - Impact: a bad spec can create a stable reconcile loop and unnecessary API/controller load.

3. Empty roles are accepted.
   - `RoleSet.spec.roles: []` is accepted and the RoleSet can become `Ready=True` with no workload Pods.
   - `StormService.spec.template.spec.roles: []` is accepted, creates an empty RoleSet, and StormService can report `replicas=1`, `readyReplicas=1` with no workload Pods.
   - Impact: this creates a false positive Ready signal.

4. Negative per-role replicas are accepted in RoleSet.
   - Observed: a role with `replicas: -1` is accepted and can become `Ready=True` with the role effectively treated as zero replicas.
   - Impact: invalid configuration is silently normalized instead of being rejected.

### Steps to reproduce

1. Deploy the latest upstream `main` CRDs and controller to minikube.
2. Create a RoleSet with duplicate role names.
3. Create a RoleSet with a role named `bad_role`.
4. Create a RoleSet with `spec.roles: []`.
5. Create a StormService whose `spec.template.spec.roles` is empty.
6. Create a RoleSet role with `replicas: -1`.
7. Check the resulting Pods, RoleSet / StormService status, events, and controller logs.

### Expected behavior

These invalid specs should be rejected during admission instead of being accepted and left for controllers to handle.

Recommended validation coverage:

- Reject empty `RoleSet.spec.roles`.
- Reject empty `StormService.spec.template.spec.roles`.
- Reject invalid role names, ideally using DNS-1123-compatible validation because role names are used to derive child resource names.
- Reject duplicate role names.
- Reject negative per-role replicas.

This can be addressed with validating webhook logic and/or CRD schema/CEL validation where appropriate. The same RoleSetSpec validation should be reused for direct RoleSet resources and nested StormService RoleSet templates.

Please also add integration tests covering these invalid specs, especially:

- Direct RoleSet admission tests.
- StormService admission tests for nested RoleSet template validation.
- Regression coverage for duplicate role names and invalid role names.

### Environment

- AIBrix version: upstream `main`, commit `136197c8db5ab353052234749a53d0534b3590de`
- Deployment environment: minikube
- Controller image: `aibrix/controller-manager:nightly`
- CRDs and controller deployed from `config/crd` and `config/default`

### PR DESCRIPTION
**Pull request by original4422**
## Pull Request Description

RoleSet role constraints were absent from the CRD schema, so invalid specs reached the controllers and could produce misleading Ready status or repeated reconciliation.

This change validates the shared RoleSetSpec at API admission:

- require at least one role;
- use the required role name as the Kubernetes map-list key, rejecting duplicates;
- require DNS-1123 role names of at most 63 characters;
- reject negative per-role replicas while preserving zero;
- require `StormService.spec.template.spec`, preventing the validating webhook from dereferencing a nil nested spec; and
- regenerate both direct RoleSet and nested StormService CRD schemas.

The integration suite now covers all invalid create cases for both resources, a missing nested StormService spec, valid zero-replica and multi-role specs, and invalid updates for both resources.

## Related Issues

Fixes #2583

## Validation

- make test-integration-webhook (60/60 passed)
- focused RoleSet admission envtest (12/12 passed)
- make test (all non-e2e/non-integration Go packages passed)
- make manifests-all generate
- make verify-codegen
- direct byte-for-byte comparison of every generated base CRD with its module and Helm chart copy

The repository verify-crd-sync script could not complete under the system Bash 3.2 because it uses associative arrays; it reported an unbound variable before comparisons. The direct comparison above covers the same generated-file consistency check.

### CODE REVIEW

**Review by googs1025 — COMMENTED** [review]
Thanks for adding the shared `RoleSetSpec` schema validation. I think the role name validation should stay: requiring DNS-1123 role names and `MaxLength=63` is useful, and the child Pod/PodSet length concern is already handled in the controller by `utils.Shorten` in `pkg/controller/roleset/utils.go` and `pkg/controller/roleset/podset_rollsyncer.go`.

One remaining edge case I think this PR should cover is `StormService.spec.template.spec == nil`. The nested `RoleSetSpec` schema only applies once `template.spec` exists, but the existing StormService validating webhook still dereferences it directly in `pkg/webhook/stormservice_webhook.go`:

```go
for _, role := range stormService.Spec.Template.Spec.Roles {
```

So a StormService missing `spec.template.spec` can still panic before the nested roles validation is reached. Could we either make `stormService.spec.template.spec` required in the CRD schema, or add an explicit nil check in the webhook that returns a validation error? A regression test for a StormService without `spec.template.spec` would be helpful too.

**Comment by original4422** [comment]
Addressed the remaining nested-spec edge case in 4a32558. `StormService.spec.template.spec` is now required by the CRD schema, and the new envtest asserts that a missing spec returns a Kubernetes `Invalid` error. Before the schema change, that assertion reproduced the validating-webhook panic; after regeneration it passes without reaching the webhook. The DNS-1123 and `MaxLength=63` role-name validation remains unchanged as suggested.

Validation: `make test-integration-webhook` (60/60), `make test`, `make verify-codegen`, and byte-for-byte CRD sync checks all passed locally.

File: test/integration/webhook/roleset_validation_test.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+			gomega.Expect(k8sClient.Create(ctx, roleSet)).To(gomega.HaveOccurred())
+		})
+
+		ginkgo.It("rejects "+tc.name+" on nested StormService creation", func() {
+			stormService := validStormService("nested-invalid", ns.Name, tc.roles)
+			gomega.Expect(k8sClient.Create(ctx, stormService)).To(gomega.HaveOccurred())
+		})
+	}
+
+	ginkgo.It("accepts zero replicas and multiple distinct roles", func() {
```
[the remark this answers is not in the document]
Comment (original4422) [reply test/integration/webhook/roleset_validation_test.go:85]: Addressed in 4a32558. The nested `spec` is now required by the CRD schema, and the regression asserts `apierrors.IsInvalid` so it distinguishes a graceful schema rejection from the previous webhook panic.

File: config/crd/orchestration/orchestration.aibrix.ai_stormservices.yaml [hunk 2]
```diff
@@ -4131,6 +4141,8 @@ spec:
                         - Sequential
                         - Interleave
                         type: string
+                    required:
+                    - roles
```
[the remark this answers is not in the document]
Comment (original4422) [reply config/crd/orchestration/orchestration.aibrix.ai_stormservices.yaml:4145]: Addressed in 4a32558 by requiring `RoleSetTemplateSpec.Spec` in the generated CRD schema. A missing nested spec is now rejected as `Invalid` before the fail-open validating webhook is invoked.

File: api/orchestration/v1alpha1/roleset_types.go [hunk 3]
```diff
@@ -195,10 +198,15 @@ type DisruptionTolerance struct {
 }
 
 type RoleSpec struct {
-	Name string `json:"name,omitempty"`
+	// Name identifies the role and is used to derive child resource names.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=63
```
[the remark this answers is not in the document]
Comment (original4422) [reply api/orchestration/v1alpha1/roleset_types.go:203]: Keeping DNS-1123 and `MaxLength=63` unchanged based on the maintainer feedback: derived child Pod/PodSet names are handled by `utils.Shorten` in the RoleSet controller.

**Review by googs1025 — COMMENTED** [review]
overall LGTM. 
Could we also add this as an integration webhook regression test under `test/integration/webhook`?

**Comment by googs1025** [comment]
For the regression test, I think it should follow the existing webhook test structure rather than adding a separate style of test. There is already a StormService validation table in `test/integration/webhook/stormservice_webhook_test.go`:

```go
ginkgo.DescribeTable("test validating", ...)
```

**Comment by original4422** [comment]
Addressed in 5e795ce. I moved the missing nested `RoleSet` spec regression into the existing `ginkgo.DescribeTable("test validating", ...)` in `test/integration/webhook/stormservice_webhook_test.go`, and removed the standalone case from `roleset_validation_test.go`. The table case still asserts `apierrors.IsInvalid(err)` so it distinguishes the intended schema rejection from a webhook panic.

Validation:
- focused StormService validation table: 6/6 specs passed
- `make test-integration-webhook`: 60/60 specs passed (including `go fmt ./...` and `go vet ./...`)
- `git diff --check`: passed

**Review by googs1025 — APPROVED** [review]
overall LGTM.

**Comment by original4422** [comment]
could this PR be merged into main?

**Comment by googs1025** [comment]
LGTM, thanks for addressing the validation gaps.
